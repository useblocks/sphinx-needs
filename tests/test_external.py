import json
from pathlib import Path

import pytest
from syrupy.filters import props


@pytest.mark.parametrize("test_app", [{"buildername": "html", "srcdir": "doc_test/external_doc"}], indirect=True)
def test_external_html(test_app):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text()
    assert (
        '<a class="external_link reference external" href="http://my_company.com/docs/v1/index.html#TEST_02">'
        "EXT_TEST_02</a>" in html
    )

    assert (
        '<p>Test need ref: <a class="external_link reference external"'
        ' href="http://my_company.com/docs/v1/index.html#TEST_01">EXT_TEST_01</a></p>' in html
    )


@pytest.mark.parametrize("test_app", [{"buildername": "needs", "srcdir": "doc_test/external_doc"}], indirect=True)
def test_external_json(test_app, snapshot):
    app = test_app
    app.build()
    json_data = Path(app.outdir, "needs.json").read_text()
    needs = json.loads(json_data)
    assert needs == snapshot(exclude=props("created"))


@pytest.mark.parametrize("test_app", [{"buildername": "needs", "srcdir": "doc_test/external_doc"}], indirect=True)
def test_external_needs_warnings(test_app):
    import os
    import subprocess

    app = test_app

    srcdir = Path(app.srcdir)
    out_dir = os.path.join(srcdir, "_build")

    out = subprocess.run(["sphinx-build", "-b", "html", srcdir, out_dir], capture_output=True)
    assert (
        "WARNING: Couldn't create need EXT_TEST_03. Reason: The need-type (i.e. `ask`) is not"
        " set in the project's 'need_types' configuration in conf.py." in out.stderr.decode("utf-8")
    )
