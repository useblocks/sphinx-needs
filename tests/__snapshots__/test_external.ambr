# serializer version: 1
# name: test_external_json[test_app0]
  dict({
    'current_version': '1.0',
    'project': 'Python',
    'versions': dict({
      '1.0': dict({
        'filters': dict({
        }),
        'filters_amount': 0,
        'needs': dict({
          'EXT_TEST_01': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': None,
            'created_at': '',
            'delete': False,
            'description': 'EXT_TEST_01',
            'docname': None,
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': 'http://my_company.com/docs/v1/index.html#TEST_01',
            'full_title': 'TEST_01 DESCRIPTION',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'EXT_TEST_01',
            'id_prefix': '',
            'is_external': True,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': False,
            'layout': None,
            'links': list([
              'SPEC_1',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': '',
            'sections': list([
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'EXT_TEST_01',
            'template': None,
            'title': 'TEST_01 DESCRIPTION',
            'type': 'impl',
            'type_name': 'Implementation',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'EXT_TEST_02': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': None,
            'created_at': '',
            'delete': False,
            'description': 'EXT_TEST_02',
            'docname': None,
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': 'http://my_company.com/docs/v1/index.html#TEST_02',
            'full_title': 'TEST_02 DESCRIPTION',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'EXT_TEST_02',
            'id_prefix': '',
            'is_external': True,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': False,
            'layout': None,
            'links': list([
              'EXT_TEST_01',
              'REQ_1',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': '',
            'sections': list([
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': 'open',
            'style': None,
            'tags': list([
              'test_02',
              'test',
            ]),
            'target_id': 'EXT_TEST_02',
            'template': None,
            'title': 'TEST_02 DESCRIPTION',
            'type': 'req',
            'type_name': 'Requirement',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'REQ_1': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'REQ_1',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Test requirement 1',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'REQ_1',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT EXTERNAL',
            'sections': list([
              'TEST DOCUMENT EXTERNAL',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'REQ_1',
            'template': None,
            'title': 'Test requirement 1',
            'type': 'req',
            'type_name': 'Requirement',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SPEC_1': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'SPEC_1',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'index',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Test specification 1',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SPEC_1',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
              'REQ_1',
              'EXT_TEST_01',
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'TEST DOCUMENT EXTERNAL',
            'sections': list([
              'TEST DOCUMENT EXTERNAL',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'SPEC_1',
            'template': None,
            'title': 'Test specification 1',
            'type': 'spec',
            'type_name': 'Specification',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
          'SUB_002': dict({
            'arch': dict({
            }),
            'avatar': '',
            'closed_at': '',
            'completion': '',
            'constraints': list([
            ]),
            'constraints_passed': True,
            'constraints_results': dict({
            }),
            'content_id': 'SUB_002',
            'created_at': '',
            'delete': None,
            'description': '',
            'docname': 'subfolder_a/subfolder_b/subpage',
            'doctype': '.rst',
            'duration': '',
            'external_css': 'external_link',
            'external_url': None,
            'full_title': 'Sub-Req',
            'has_dead_links': '',
            'has_forbidden_dead_links': '',
            'hidden': '',
            'id': 'SUB_002',
            'id_prefix': '',
            'is_external': False,
            'is_modified': False,
            'is_need': True,
            'is_part': False,
            'jinja_content': None,
            'layout': '',
            'links': list([
            ]),
            'max_amount': '',
            'max_content_lines': '',
            'modifications': 0,
            'params': '',
            'parent_need': '',
            'parent_needs': list([
            ]),
            'parts': dict({
            }),
            'post_template': None,
            'pre_template': None,
            'prefix': '',
            'query': '',
            'section_name': 'Subpage',
            'sections': list([
              'Subpage',
            ]),
            'service': '',
            'signature': '',
            'specific': '',
            'status': None,
            'style': None,
            'tags': list([
            ]),
            'target_id': 'SUB_002',
            'template': None,
            'title': 'Sub-Req',
            'type': 'req',
            'type_name': 'Requirement',
            'updated_at': '',
            'url': '',
            'url_postfix': '',
            'user': '',
          }),
        }),
        'needs_amount': 5,
      }),
    }),
  })
# ---
